docs/RFCS/time_travel.md

```diff
@@ -0,0 +1,81 @@
+- Feature Name: Time Travel Queries
+- Status: draft
+- Start Date: 2016-05-07
+- Authors: Matt Jibson
+- RFC PR: #
+- Issue PR: [#5963](https://github.com/cockroachdb/cockroach/issues/5963)
+
+# Summary
+
+Add support for `AS OF SYSTEM TIME` to `SELECT` statements to execute
+queries at some past time. Plumb doing things at a timestamp to various
+layers that previously did not need it.
+
+# Motivation
+
+[SQL:2011](https://en.wikipedia.org/wiki/SQL:2011) adds support for time
+travel queries. Our most immediate need for them is to support backups
+at a certain timestamp. (Consider a huge table being backed up. We need
+to page through the data in a way that will be consistent if a connection
+drops and needs to be resumed from where it left off.)
+
+The syntax is: `SELECT * FROM t AS OF SYSTEM TIME '2016-01-01 00:00'`.
+The term `TIME` will be used below as the time travel time.
+
+# Detailed design
+
+SQL queries already create a KV transaction with a timestamp. Hooking
+up the `AS OF SYSTEM TIME` syntax to set the transaction timestamp is
+straightforward. This causes older data in the MVCC table to be fetched.
+
+There are other issues that need to be taken into account.
+
+## Table Schema
+
+Before a SELECT, a table lease is obtained to determine the schema. If the
+schema has changed after `TIME`, we need to instead fetch the old schema
+as well. Since the schemas are stored in the KV store, we can use the same
+technique (setting the KV transaction timestamp) to fetch the old data.
+
+Since this query is known to be readonly, it should not trigger a lease even
+if `TIME` is now and the schema is current. We just need the descriptor at
+`TIME`.
+
+## MVCC Garbage Collection
+
+MVCC currently garbage collects old data after 1 day. If data is requested
+at a time before the oldest value, the MVCC layer will return empty data
+(not an error). This will be changed to return an error instead of an
+empty value. A marker will be added to record the earliest time at which
+the oldest value is valid.
+
+TODO: is there a better way than the marker? Is there something already
+that I’m not aware of?
+
+TODO: can we throw an error at the SQL layer (i.e., don’t query KV at all)
+if `TIME` is more than the GC time in the past?
+
```

Comment:
Perhaps via a statement that temporarily suspends garbage collection? Or disable garbage collection when there are currently some time travel queries being processed? (This is dangerous, you may not want to resume GC if the backup connection drops unexpectedly)